codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res

```diff
@@ -186,89 +214,146 @@ let setChainFetcher = (self: t, chainFetcher: ChainFetcher.t) => {
   //Compare the peeked items and determine the next item
   switch fetchStatesMap->determineNextEvent(~isUnorderedMultichainMode) {
-  | Ok({chain, earliestEvent}) =>
+  | Ok({val: {chain, earliestEvent}, isInReorgThreshold}) =>
     let maybeArbItem = if isUnorderedMultichainMode {
-      arbitraryEventQueue->getFirstArbitraryEventsItemForChain(~chain)
+      arbitraryEventQueue->getFirstArbitraryEventsItemForChain(~chain, ~fetchStatesMap)
     } else {
-      arbitraryEventQueue->getFirstArbitraryEventsItem
+      arbitraryEventQueue->getFirstArbitraryEventsItem(~fetchStatesMap)
     }
     switch maybeArbItem {
     //If there is item on the arbitray events queue, and it is earlier than
     //than the earlist event, take the item off from there
-    | Some(itemWithPopFn)
+    | Some({val: itemWithPopFn, isInReorgThreshold})
       if Item(itemWithPopFn)->getQueueItemComparitor(~chain=itemWithPopFn.item.chain) <
-        earliestEvent->getQueueItemComparitor(~chain) =>
-      Some(itemWithPopFn)
+        earliestEvent->getQueueItemComparitor(~chain) => {
+        isInReorgThreshold,
+        val: Some(itemWithPopFn),
+      }
     | _ =>
       switch earliestEvent {
-      | NoItem(_) => None
-      | Item(itemWithPopFn) => Some(itemWithPopFn)
+      | NoItem(_) => {
+          isInReorgThreshold,
+          val: None,
+        }
+      | Item(itemWithPopFn) => {
+          isInReorgThreshold,
+          val: Some(itemWithPopFn),
+        }
       }
     }
-  | Error(NoItemsInArray) => arbitraryEventQueue->getFirstArbitraryEventsItem
+  | Error(NoItemsInArray) =>
+    arbitraryEventQueue
+    ->getFirstArbitraryEventsItem(~fetchStatesMap)
+    ->Option.mapWithDefault({val: None, isInReorgThreshold: false}, ({val, isInReorgThreshold}) => {
+      isInReorgThreshold,
+      val: Some(val),
+    })
   }
 }
 
+let getFetchStateWithData = (self: t, ~shouldDeepCopy=false): ChainMap.t<fetchStateWithData> => {
+  self.chainFetchers->ChainMap.map(cf => {
+    partitionedFetchState: shouldDeepCopy
+      ? cf.fetchState->PartitionedFetchState.copy
+      : cf.fetchState,
+    heighestBlockBelowThreshold: cf.currentBlockHeight - cf.chainConfig.confirmedBlockThreshold,
+  })
+}
+
 /**
 Simply calls popBatchItem in isolation using the chain manager without
 the context of a batch
 */
 let nextItemIsNone = (self: t): bool => {
   popBatchItem(
-    ~fetchStatesMap=self.chainFetchers->ChainMap.map(cf => cf.fetchState),
+    ~fetchStatesMap=self->getFetchStateWithData,
     ~arbitraryEventQueue=self.arbitraryEventQueue,
     ~isUnorderedMultichainMode=self.isUnorderedMultichainMode,
-  )->Option.isNone
+  ).val->Option.isNone
 }
 
-type batchRes = {
-  batch: array<Types.eventBatchQueueItem>,
-  fetchStatesMap: ChainMap.t<PartitionedFetchState.t>,
-  arbitraryEventQueue: array<Types.eventBatchQueueItem>,
+let hasChainItemsOnArbQueue = (self: t, ~chain): bool => {
+  self.arbitraryEventQueue->Js.Array2.find(item => item.chain == chain)->Option.isSome
```

Comment:
Just for you to know, there's https://rescript-lang.org/docs/manual/latest/api/js/array2#value-some